Exporting package metadata as json

giacomocavalieri opened the pull request:
This PR closes #1475

### ~~Why is this a draft PR?~~
- [x] I still have to add a bunch of tests to make sure everything works as expected (maybe snapshot tests with insta? Sounds like the best way to do it)
- [x] I still have to understand when this file should be generated and where to put it (should this only be enabled with a flag? Or be always generated together with the html documentation? Should I be able to choose the destination path with a CLI argument?)
- [x] The target tracking should change to take into account pure Gleam functions supported by _all_ modules (even future ones) and those that support all _current_ modules through external implementations
- [x] There's a bug where package documentation is not included in the interface
- [x] I still have to add some documentation

giacomocavalieri commented:
I think the PR is now ready for review, there's a [point in the tests I added](https://github.com/gleam-lang/gleam/pull/2537/files#diff-cba801b5bd04aeee35b67c4cbb7114e5259da681de23c9e589dedce95d30b84aR32-R80) that I think could be improved: it's a bit of code that is copy pasted and exactly the same for the `erlang` end `javascript` tests as well.
I could make a single function to avoid the code duplication, but I'm not really sure where to put such function

lpil commented on compiler-cli/src/main.rs:
```diff
@@ -231,15 +231,22 @@ enum Command {
+    /// The package json interface
+    PackageInterface {
+        #[clap(long = "out", required = true)]
+        /// A file to write the json interface to, relative to the package root
```
```suggestion
        /// The path to write the JSON file to
```
Might not be relative

lpil commented on compiler-cli/src/main.rs:
```diff
@@ -231,15 +231,22 @@ enum Command {
     JavascriptPrelude,
+    /// The TypeScript prelude module
     TypescriptPrelude,
+    /// The package json interface
```
```suggestion
    /// Information on the modules, functions, and types in the project in JSON format
```

lpil commented on compiler-core/src/manifest.rs:
```diff
@@ -193,7 +193,7 @@ pub enum ManifestPackageSource {
 }
 
-fn ordered_map<S, K, V>(value: &HashMap<K, V>, serializer: S) -> Result<S::Ok, S::Error>
+pub fn ordered_map<S, K, V>(value: &HashMap<K, V>, serializer: S) -> Result<S::Ok, S::Error>
```
If this is public it should not be in the manifest module as it's unrelated to the manifest.

giacomocavalieri replied:
What would be a good place to put this function?

lpil replied:
It's annoyingly generic isn't it? Perhaps `io` at a push

lpil commented on compiler-cli/src/main.rs:
```diff
@@ -231,15 +231,22 @@ enum Command {
+    PackageInterface {
+        #[clap(long = "out", required = true)]
+        /// A file to write the json interface to, relative to the package root
+        output: String,
```
```suggestion
        output: Utf8PathBuf,
```

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+        #[serde(rename = "return")]
+        return_: Box<TypeInterface>,
+    },
+    /// A generic type variable.
```
```suggestion
    /// A type variable.
```

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+
+            // Type aliases are ignored since those are also erased from the
+            // generated documentation.
+            Definition::TypeAlias(_) => {}
```
Not true, they can have the pub keyword!

giacomocavalieri replied:
Ok, I'll include type aliases as well!

lpil commented on compiler-core/src/package_interface/tests.rs:
```diff
@@ -0,0 +1,201 @@
+
+    // TODO: all the bits above are basically copy pasted from the javascript
+    // and erlang test helpers. A refactor might be due here, but I'm not sure
+    // where to put such function...
```
Yeah we really need something that does all this. So much duplication.

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+                // Package serialisation happens after inference has taken place
+                // At this point no one should be helding a mutable reference of
+                // the type: the borrow should never fail.
+                Err(_) => unreachable!(),
```
Use `expect` here, that's how we make assertions about things that must hold true and otherwise be a fatal error.

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+/// Holds serialisable data about a top-level public value defined in a module
+/// (either a constant or a function)
+#[derive(Serialize, Debug)]
+pub struct ValueInterface {
```
Could you include the supported targets too please 🙏 

Need to decide what's a good encoding for "all". Could be a null value means everything, or it could be that there's a "gleam" target which means all (`["gleam", "erlang"]`). Or something else.

giacomocavalieri replied:
There's a `supported_targets` field in this struct; it is an array of targets so "all" would just become a list with all the targets like this: `["erlang", "javascript"]`

lpil replied:
Oh yes! Sorry I missed that.

It needs to distinguish between there being an external implementation for both and for there being a Gleam implementation as otherwise if a wasm backend arrives then we can no longer tell which are supported, and all packages will need to be republished. I think we'd need to extend the target tracking to have this level of information.

giacomocavalieri replied:
Aha! You're right it'd be quite a problem if we don't track that... I'll have to put this on hold and update the target tracking code first, It shouldn't be too hard

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+                    name.clone(),
+                    ValueInterface {
+                        supported_targets: supported_targets.to_vec(),
+                        deprecation: deprecation.clone(),
```
This should be null if it's not deprecated, not the string "NotDeprecated"

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,280 @@
+                    //       overlap.
+                    //       Once #2533 is closed this branch can be turned back to
+                    //       be unreachable!().
+                    TypeVar::Unbound { id } => TypeInterface::Variable { id: *id },
+                    TypeVar::Link { type_ } => TypeInterface::from_type(type_),
+                    TypeVar::Generic { id } => TypeInterface::Variable { id: *id },
```
This should use new ints for each value, not the internal ints from the compiler. So if a function's type has 3 variables they will always be 0, 1, and 2.

lpil reviewed (commented):
Amazing! Thank you. I've left some notes

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,596 @@
+pub struct PackageInterface {
+    name: EcoString,
+    package_version: Version,
+    /// The minimum Gleam version required by the package.
+    gleam_version: Option<EcoString>,
```
```suggestion
    /// The Gleam version constraint that the package specifies in its `gleam.toml`
    gleam_version_constraint: Option<EcoString>,
```

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,596 @@
+    Fn {
+        parameters: Vec<ParameterInterface>,
+        #[serde(rename = "return")]
+        return_: Box<TypeInterface>,
```
Labels are not part of the type system so this information cannot be part of the type AST.

The only places one can have a label is on a module function or on a record constructor so I think we'll need to have fields on those that hold this information.

lpil reviewed (commented):
Looking really good! Some small comments

giacomocavalieri commented:
I've addressed the review comments and did some little changes:
- I have [removed the duplicate step in the CI](https://github.com/gleam-lang/gleam/pull/2537/commits/db837c095239e5e14a31d543c2b263c96156257f) as you suggested
- I have changed the `gleam_version` to `gleam_version_constraint` and updated its documentation as you suggested. I've also [changed the `Version` type to `String`](https://github.com/gleam-lang/gleam/pull/2537/commits/d47780114b8561ef60ff8d564e2a5cf165a53403) since the its `Serialize` implementation already turns it into a string and this makes it easier to see it without having to go look for the `Serialize` implementation
- I have defined a `ImplementationsInterface` that replaces `Implementations`, you can read more about my reasoning in the [comments I've added](https://github.com/gleam-lang/gleam/pull/2537/commits/c8e8382e6d18610507aed8ef20b5b4c67a9dc6ca)
- Like you suggested I have [removed labels from the definition of a function type](https://github.com/gleam-lang/gleam/pull/2537/commits/85d96f96dadd8ad61d712f572ecc551735b45405). This also led to updating the module interface to have two listing fields: `constants` and `functions`. And module `functions` now have labelled arguments

lpil commented:
Q: Why are the version and the Gleam constraint strings rather then their normal types? Wouldn't it be better to use the precise types to ensure they are valid? So far the compiler avoids using strings instead of precise types whenever possible.

giacomocavalieri commented:
> Q: Why are the version and the Gleam constraint strings rather then their normal types? Wouldn't it be better to use the precise types to ensure they are valid? So far the compiler avoids using strings instead of dedicated types whenever possible.

As for `gleam_version_constraint` I went with a `String` because the compiler itself is using a string and [I don't have access to anything more specific in the Package's `PackageConfig`.](https://github.com/gleam-lang/gleam/pull/2537/files#diff-713f951c0b7f4c757dd5cccc55c5cd1ad26d2ff609cd2cc4cc961e92e9a628a3R331)

As for the `version` field I used a string since that's what it gets serialised into and it makes it easier to see. I agree that it would be better to use a more specific structure though:
```rust
#[derive(Serialize)]
struct VersionInterface {
  major: u32,
  minor: u32,
  patch: u32,
  pre: Vec<PreIdentifier>,
  build: Option<String>,
}

#[derive(Serialize)]
enum PreIdentifier {
  Numeric(u32),
  AlphaNumeric(String),
}
```

lpil commented on CHANGELOG.md:
```diff
@@ -7,19 +7,24 @@
 
 ### Build tool
 
+- The search bar in generated docs now has a darker background color.
```
```suggestion
```

lpil commented on compiler-core/src/package_interface/snapshots/gleam_core__package_interface__tests__generic_function.snap:
```diff
@@ -0,0 +1,71 @@
+---
+{
+  "name": "my_package",
+  "package_version": {
```
This should be `"version": "11.10.9-1foo"`. I think the default encoder for a version uses that format so you could try that.

lpil commented on compiler-core/src/package_interface.rs:
```diff
@@ -0,0 +1,680 @@
+#[derive(Serialize, Debug)]
+pub struct PackageInterface {
+    name: EcoString,
+    package_version: VersionInterface,
```
```suggestion
    version: VersionInterface,
```

lpil reviewed (commented):
Could you have the JSON keys use `-` rather than `_` for word delimiters please? As on the web `_` means it's the one word, while `-` means it's two words. Thank you.

giacomocavalieri commented:
> Could you have the JSON keys use - rather than _ for word delimiters please? As on the web _ means it's the one word, while - means it's two words. Thank you.

Thanks for the review, I wasn't aware of that! Now it should be fixed :)

lpil reviewed (approved):
Thank you so much!! 🎉
